rfcs/20190815-tfdbg-v2-callbacks.md

```diff
@@ -0,0 +1,356 @@
+        #          If the current invocation of the callback is due to the
+        #          eager execution of an op, this will be `None` as op names
+        #          are meaningless in eager execution.
+        #          If this invocation of the callback is due to the
+        #          eager execution of a FuncGraph, this will be the
+        #          internally-generated name of the FuncGraph.
+        #          In graph construction, this is the name of the op.
+        # graph: The graph that the op belongs to (if any).
+        #        In eager execution of an op, this is `None`.
+        #        In eager execution of a FuncGraph, the is the FuncGraph
+        #        object per se.
+        #        In graph construction, this is the op's containing graph.
+        #
+        # Return values:
+        #   This callback function is expected to return a `list` or `tuple`
+        #   of `Tensor`s, with its length matching `len(outputs)`.
+        #   In eager execution, these returned `Tensor`s should be
+        #   `EagerTensor`s. Their values will replace the original values of
+        #   `outputs` during downstream eager execution.
+        #   In graph construction, these returned `Tensor`s should be
+        #   non-eager `Tensor`s. Their values will replace the original
+        #   `outputs` during downstream graph construction.
+
+  Returns:
+    A thread-local context manager. Within the scope of the context
+    manager, all eager op/graph execution and graph op construction
+    will invoke `callback_fn`.
+
+  Raises:
+    ValueEror: If the `callback` does not return the a `list` or `tuple`
+    of `Tensor`s with the same length as the `outputs` argument passed to it.
+  """
+```
+
+This API follows the style of TF2's op API, namely a style that **unifies eager
+and graph modes**. In TF2, the same op API (e.g., `tf.matmul`) executes
+differently depending on where it is called. If it is called in an eager
+context, it will execute eagerly, consuming `EagerTensor`s and generating new
+`EagerTensor`s as outputs. If called in a graph context (i.e., in
+`tf-function`), it'll construct a new symbolic op by consuming inbound symbolic
+Tensors. Our proposal captures both cases with a single API.
+
+The proposed API achieves Capabilities A and B listed above, due to the fact
+that the callback(s) registered with the API will be invoked for both eager op
+execution and graph op creation.
+
+Capability C is also met by this API, due to the fact that the callbacks are
+invoked not only during the eager execution of ops, but also the execution
+of FuncGraphs. In this paradigm, when a FuncGraph has been constructed and
+undergoes execution by TF2, it is simply treated as a special type of op.
+
+The example code snippet below shows how Capability A is met, i.e., how eager
+execution of ops is intercepted by the callback mechanism.
+
+```python
+with tf.debugging.op_callback(my_callback):
```

Comment:
Can you include what my_callback is in this case? This is written by the user?

Reply:
Added further clarification in the code comment.